File: configd/src/apps/sentinel/rpcserver.cpp
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.

#include "rpcserver.h"

#include <vespa/log/log.h>
LOG_SETUP(".sentinel.rpcserver");

namespace config::sentinel {

RpcServer::RpcServer(int portNumber, CommandQueue &cmdQ, ModelOwner &modelOwner)
    : _server(),
      _rpcHooks(cmdQ, _server.supervisor(), modelOwner),
      _port(portNumber)
{
    if (_server.supervisor().Listen(portNumber)) {
        LOG(config, "listening on port %d", portNumber);
    } else {
        LOG(error, "unable to listen to port %d", portNumber);
    }
}

RpcServer::~RpcServer() = default;

} // namespace config::sentinel

File: logserver/src/main/java/ai/vespa/logserver/protocol/RpcServer.java
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package ai.vespa.logserver.protocol;

import com.yahoo.jrt.Acceptor;
import com.yahoo.jrt.ListenFailedException;
import com.yahoo.jrt.Method;
import com.yahoo.jrt.Spec;
import com.yahoo.jrt.Supervisor;
import com.yahoo.jrt.Transport;

/**
 * A JRT based RPC server for handling log requests
 *
 * @author bjorncs
 */
public class RpcServer implements AutoCloseable {

    private final Supervisor supervisor;
    private final int listenPort;
    private Acceptor acceptor;

    public RpcServer(int listenPort) {
        supervisor = new Supervisor(new Transport("logserver-" + listenPort));
        this.listenPort = listenPort;
    }

    public void addMethod(Method method) {
        supervisor.addMethod(method);
    }

    public void start() {
        try {
            acceptor = supervisor.listen(new Spec(listenPort));
        } catch (ListenFailedException e) {
            throw new RuntimeException(e);
        }
    }

    int listenPort() {
        return acceptor.port();
    }

    @Override
    public void close() {
        if (acceptor != null) {
            acceptor.shutdown().join();
        }
        supervisor.transport().shutdown().join();
    }
}
